File: public/public-persistence-jooq/src/main/java/ch/difty/scipamato/publ/persistence/newstudies/JooqNewStudyTopicService.java
package ch.difty.scipamato.publ.persistence.newstudies;

import java.util.Collections;
import java.util.List;

import org.jetbrains.annotations.NotNull;
import org.springframework.stereotype.Service;

import ch.difty.scipamato.publ.entity.NewStudyPageLink;
import ch.difty.scipamato.publ.entity.NewStudyTopic;
import ch.difty.scipamato.publ.entity.Newsletter;
import ch.difty.scipamato.publ.persistence.api.NewStudyTopicService;

/**
 * Currently only stubbed. Implement from table
 *
 * @author Urs Joss
 */
@Service
public class JooqNewStudyTopicService implements NewStudyTopicService {

    private final NewStudyRepository repo;

    public JooqNewStudyTopicService(@NotNull final NewStudyRepository repo) {
        this.repo = repo;
    }

    @NotNull
    @Override
    public List<NewStudyTopic> findMostRecentNewStudyTopics(@NotNull final String languageCode) {
        return repo
            .findMostRecentNewsletterId()
            .map(id -> repo.findNewStudyTopicsForNewsletter(id, languageCode))
            .orElse(Collections.emptyList());
    }

    @NotNull
    @Override
    public List<NewStudyTopic> findNewStudyTopicsForNewsletterIssue(@NotNull final String issue,
        @NotNull final String languageCode) {
        return repo
            .findIdOfNewsletterWithIssue(issue)
            .map(id -> repo.findNewStudyTopicsForNewsletter(id, languageCode))
            .orElse(Collections.emptyList());
    }

    @NotNull
    @Override
    public List<Newsletter> findArchivedNewsletters(final int newsletterCount, @NotNull final String languageCode) {
        return repo.findArchivedNewsletters(newsletterCount, languageCode);
    }

    @NotNull
    @Override
    public List<NewStudyPageLink> findNewStudyPageLinks(@NotNull final String languageCode) {
        return repo.findNewStudyPageLinks(languageCode);
    }
}

File: public/public-persistence-jooq/src/main/kotlin/ch/difty/scipamato/publ/persistence/newstudies/JooqNewStudyTopicService.kt
package ch.difty.scipamato.publ.persistence.newstudies

import ch.difty.scipamato.publ.entity.NewStudyPageLink
import ch.difty.scipamato.publ.entity.NewStudyTopic
import ch.difty.scipamato.publ.entity.Newsletter
import ch.difty.scipamato.publ.persistence.api.NewStudyTopicService
import org.springframework.stereotype.Service

/**
 * Currently only stubbed. Implement from table
 */
@Service
class JooqNewStudyTopicService(private val repo: NewStudyRepository) : NewStudyTopicService {

    override fun findMostRecentNewStudyTopics(languageCode: String): List<NewStudyTopic> =
        repo.findMostRecentNewsletterId()?.let {
            repo.findNewStudyTopicsForNewsletter(it, languageCode)
        } ?: emptyList()

    override fun findNewStudyTopicsForNewsletterIssue(
        issue: String,
        languageCode: String,
    ): List<NewStudyTopic> = repo.findIdOfNewsletterWithIssue(issue)?.let {
        repo.findNewStudyTopicsForNewsletter(it, languageCode)
    } ?: emptyList()

    override fun findArchivedNewsletters(newsletterCount: Int, languageCode: String): List<Newsletter> =
        repo.findArchivedNewsletters(newsletterCount, languageCode)

    override fun findNewStudyPageLinks(languageCode: String): List<NewStudyPageLink> =
        repo.findNewStudyPageLinks(languageCode)
}
